rfcs/2020-12-08-5457-dynamic-sampling.md

```diff
@@ -0,0 +1,355 @@
+# RFC 5457 - <2020-12-08> - Dynamic sampling in the `sampler` transform
+
+The addition of **dynamic sampling** capabilities (**DS** henceforth) to Vector would enable us to
+cover a much broader range of sampling use cases than we currently can. A qualitatively better
+sampling story would enhance Vector's usefulness in the logging domain while also potentially
+putting Vector in a better position to move into domains like distributed tracing.
+
+## Scope
+
+This RFC proposes extensions to the [`sampler`][sampler] transform, including new configuration
+parameters and the introduction of the **bucket** concept to Vector.
+
+### Goals and non-goals
+
+The purpose of the RFC is to achieve consensus around a set of desired features. Thus, the RFC does
+*not* cover:
+
+* Any Vector internals that would affect other components, e.g. the processing pipeline or
+  configuration system
+* Implementation details on the Rust side
+* Finalized names for configuration parameters
+* The specifics of sampling algorithms; consensus around that should be deferred to the development
+  process or another RFC
+
+### Domains
+
+Dynamic sampling has implications for both logging and tracing. It does *not* have implications for
+metrics, as metrics are already a form of sampling in themselves, e.g. counters reveal a raw number
+of events while saying little about the content of those events.
+
+This RFC, however, only covers logging and not tracing. It's possible that tracing would be
+seamlessly covered by the changes proposed here but I lack the domain-specific knowledge to assert
+that, so I will leave it as an open question for a future RFC.
```

Comment:
Agree. Sampling traces is much much more complicated. Look at the Lightstep agent for prior art.